packages/near-membrane-base/src/membrane.ts

```diff
@@ -455,64 +454,51 @@ export function createMembraneMarshall() {
 
         function getTargetTraits(target: object): TargetTraits {
             let targetTraits = TargetTraits.None;
+            let targetIsArray = false;
+            try {
+                targetIsArray = isArrayOrNotOrThrowForRevoked(target);
+            } catch {
+                targetTraits |= TargetTraits.Revoked;
+            }
             if (typeof target === 'function') {
                 targetTraits |= TargetTraits.IsFunction;
-                // detecting arrow function vs function
+                // A target may be a proxy that throws in its "has" trap.
                 try {
+                    // Detect arrow functions.
                     targetTraits |= +!('prototype' in target) && TargetTraits.IsArrowFunction;
-                } catch {
-                    // target is either a revoked proxy, or a proxy that
-                    // throws on the `has` trap, in which case going with
-                    // a strict mode function seems appropriate.
-                }
+                    // eslint-disable-next-line no-empty
+                } catch {}
+            } else if (targetIsArray) {
+                targetTraits |= TargetTraits.IsArray;
             } else {
-                let targetIsArray = false;
-                try {
-                    // try/catch in case Array.isArray throws when target
-                    // is a revoked proxy
-                    targetIsArray = isArrayOrNotOrThrowForRevoked(target);
-                } catch {
-                    // TODO: this might be problematic, because functions
-                    // and arrow functions should also be subject to this,
-                    // but it seems that we can still create a proxy of
-                    // a revoke, and wait until the user-land code actually
-                    // access something out of it to throw the proper error.
-                    // target is a revoked proxy, so the type doesn't matter
-                    // much from this point on
-                    targetTraits |= TargetTraits.Revoked;
-                }
-                targetTraits |= +targetIsArray && TargetTraits.IsArray;
-                targetTraits |= +!targetIsArray && TargetTraits.IsObject;
+                targetTraits |= TargetTraits.IsObject;
```

Comment:
does this mean the traits will have isArray and isObject for any array?

Reply:
I'm just curious, I don't see any problem with it.

Reply:
> does this mean the traits will have isArray and isObject for any array?

No. The logic here is unchanged in that before if a value was an array it was marked as an array and not an object. I've just made it more explicit in the if/else to avoid the `targetTraits |= +!targetIsArray && TargetTraits.IsObject;` bit.